### PR TITLE
chore: pipeline dryrun api can currently receives pipeline raw content

### PR DESCRIPTION
**Pull request by paomian**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

* pipeline dryrun api can currently receives pipeline raw content

### dryrun pipeline via specify pipeline_name

```
curl -X "POST" "http://localhost:4000/v1/events/pipelines/v1/dryrun?db=public" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "pipeline_name": "apach_test",
  "data": [
    "145.251.205.39 - devankoshal [29/Oct/2024:17:51:29 +0800] \\"DELETE /booper/bopper/mooper/mopper HTTP/1.1\\" 304 43420"
  ]
}'

{"rows":[[{"data_type":"STRING","key":"ip_address","semantic_type":"TAG","value":"145.251.205.39"},{"data_type":"STRING","key":"http_method","semantic_type":"TAG","value":"DELETE"},{"data_type":"STRING","key":"http_version","semantic_type":"TAG","value":"1.1"},{"data_type":"INT32","key":"status_code","semantic_type":"TAG","value":304},{"data_type":"STRING","key":"request_path","semantic_type":"FIELD","value":"/booper/bopper/mooper/mopper"},{"data_type":"STRING","key":"user","semantic_type":"FIELD","value":"devankoshal"},{"data_type":"INT32","key":"response_size","semantic_type":"FIELD","value":43420},{"data_type":"TIMESTAMP_NANOSECOND","key":"request_time","semantic_type":"TIMESTAMP","value":"2024-10-29 09:51:29+0000"}]],"schema":[{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"ip_address"},{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"http_method"},{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"http_version"},{"colume_type":"TAG","data_type":"INT32","fulltext":false,"name":"status_code"},{"colume_type":"FIELD","data_type":"STRING","fulltext":true,"name":"request_path"},{"colume_type":"FIELD","data_type":"STRING","fulltext":false,"name":"user"},{"colume_type":"FIELD","data_type":"INT32","fulltext":false,"name":"response_size"},{"colume_type":"TIMESTAMP","data_type":"TIMESTAMP_NANOSECOND","fulltext":false,"name":"request_time"}]}
```

### dryrun pipeline via specify pipeline raw content

```
curl -X "POST" "http://localhost:4000/v1/events/pipelines/v1/dryrun" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "pipeline": "processors:\\n  - dissect:\\n      fields:\\n        - message\\n      patterns:\\n        - \'%{ip_address} - %{user} [%{request_time}] \\"%{http_method} %{request_path} HTTP/%{http_version}\\" %{status_code} %{response_size}\'\\n      ignore_missing: true\\n  - date:\\n      fields:\\n        - request_time\\n      formats:\\n        # 29/Oct/2024:17:51:29 +0800\\n        - \\"%d/%b/%Y:%H:%M:%S %z\\"\\n\\ntransform:\\n  - fields:\\n      - ip_address\\n      - http_method\\n      - http_version\\n    type: string\\n    index: tag\\n  - fields:\\n      - status_code\\n    type: int32\\n    index: tag\\n  - fields:\\n      - request_path\\n    type: string\\n    index: fulltext\\n  - fields:\\n      - user\\n    type: string\\n  - fields:\\n      - response_size\\n    type: int32\\n  - fields:\\n      - request_time\\n    type: time\\n    index: timestamp",
  "data": [
    "145.251.205.39 - devankoshal [29/Oct/2024:17:51:29 +0800] \\"DELETE /booper/bopper/mooper/mopper HTTP/1.1\\" 304 43420"
  ]
}'

{"rows":[[{"data_type":"STRING","key":"ip_address","semantic_type":"TAG","value":"145.251.205.39"},{"data_type":"STRING","key":"http_method","semantic_type":"TAG","value":"DELETE"},{"data_type":"STRING","key":"http_version","semantic_type":"TAG","value":"1.1"},{"data_type":"INT32","key":"status_code","semantic_type":"TAG","value":304},{"data_type":"STRING","key":"request_path","semantic_type":"FIELD","value":"/booper/bopper/mooper/mopper"},{"data_type":"STRING","key":"user","semantic_type":"FIELD","value":"devankoshal"},{"data_type":"INT32","key":"response_size","semantic_type":"FIELD","value":43420},{"data_type":"TIMESTAMP_NANOSECOND","key":"request_time","semantic_type":"TIMESTAMP","value":"2024-10-29 09:51:29+0000"}]],"schema":[{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"ip_address"},{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"http_method"},{"colume_type":"TAG","data_type":"STRING","fulltext":false,"name":"http_version"},{"colume_type":"TAG","data_type":"INT32","fulltext":false,"name":"status_code"},{"colume_type":"FIELD","data_type":"STRING","fulltext":true,"name":"request_path"},{"colume_type":"FIELD","data_type":"STRING","fulltext":false,"name":"user"},{"colume_type":"FIELD","data_type":"INT32","fulltext":false,"name":"response_size"},{"colume_type":"TIMESTAMP","data_type":"TIMESTAMP_NANOSECOND","fulltext":false,"name":"request_time"}]}
```

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

**Comment by sunng87** [comment]
I think we can just remove the previous api because the main branch is already targeted for next minor release. An alternative option is to support both in the previous one, by checking input parameters

**Comment by paomian** [comment]
> I think we can just remove the previous api because the main branch is already targeted for next minor release. An alternative option is to support both in the previous one, by checking input parameters

Adding a new interface leaves time for changes to the logview to be made, otherwise the merged nightly version of the logview will not be available.

One problem with determining the content of the http is that if the user data is in the same format as our new dryrun json, and the pipeline_name is set in the query params, we can't tell if the body is the user's data or the new version of the dryrun json.

**Comment by sunng87** [comment]
> we can't tell if the body is the user's data or the new version of the dryrun json.

We can define a rule for that. If pipeline is provided in json, we will always use it. I think it also makes sense in terms of semantics.

**Review by sunng87 — COMMENTED** [review]
Just some suggestions on modelling query parameters:

1. don't put `pipename_name` into `pipeline_info`. we can provided a top-level key `pipeline_name` for user to specify the name
2. consider rename `pipeline_info` to `pipeline` directly
3. add a test when input contains pipeline instead of pipeline_name

**Comment by paomian** [comment]
> Just some suggestions on modelling query parameters:

What you're trying to convey is the modeling of the body not "query parameters", right?

**Comment by sunng87** [comment]
> What you're trying to convey is the modeling of the body not "query parameters", right?

Yes. We keep the query parameters as it is.

**Review by sunng87 — APPROVED** [review]
LGTM

File: src/servers/src/http/event.rs [hunk 1]
```diff
@@ -387,6 +359,89 @@ pub async fn pipeline_dryrun(
...
+    TypedHeader(content_type): TypedHeader<ContentType>,
+    payload: String,
+) -> Result<Response> {
+    let handler = log_state.log_handler;
+    fn check_pipeline_name_exists(pipeline_name: Option<String>) -> Result<String> {
+        pipeline_name.context(InvalidParameterSnafu {
+            reason: "pipeline_name is required",
+        })
+    }
+    match serde_json::from_str::<PipelineDryrunParams>(&payload) {
```
Comment (shuiyisong) [inline src/servers/src/http/event.rs:389]: The matching arms of this match are confusing, why is `Err(_) | Ok(_)` exist on the same path. I suggest refactoring this logic according to each scenario and please add more comments on each case.

Note: the description of this pr seems out of date, please update it too.

Comment (paomian) [reply src/servers/src/http/event.rs]: updated the description.
And a separate function is extracted to determine whether the new logic is used or not.

File: src/servers/src/http/event.rs [hunk 2]
```diff
@@ -387,6 +359,107 @@ pub async fn pipeline_dryrun(
...
+                        .await?;
+                    dryrun_pipeline_inner(data, &pipeline)
+                }
+                Some(pipeline) => {
+                    let pipeline = handler.build_pipeline(&pipeline)?;
+                    dryrun_pipeline_inner(data, &pipeline)
+                }
+            }
+        }
+        None => {
```
Comment (shuiyisong) [inline src/servers/src/http/event.rs:439]:
```Suggestion
        None => {
        // This path is for back compatibility with the previous dry run code
        // where the payload is just data (JSON or plain text) and the pipeline name 
        // is specified using query param.
```

**Review by shuiyisong — APPROVED** [review]
Rest LGTM

**Pull request merged**